packages/libs/deer/src/error/hooks.rs

```diff
@@ -0,0 +1,754 @@
+            ]);
+        });
+    }
+
+    fn push(&self, hooks: &[Hook]) {
+        // TODO: we need to check if the combination of Namespace and Id already exists, if that is
+        //  the case panic?
+        #[cfg(feature = "std")]
+        let mut inner = self.inner.write().expect("lock has not been poisoned");
+        #[cfg(not(feature = "std"))]
+        let mut inner = self.inner.write();
+        let inner = inner.get_or_insert(BTreeMap::new());
+
+        for hook in hooks {
+            inner.insert(hook.id, hook.hook);
+        }
+    }
+
+    fn read_hooks_with<T>(&self, f: impl FnOnce(Option<&BTreeMap<TypeId, HookFn>>) -> T) -> T {
+        #[cfg(feature = "std")]
+        let hooks = self.inner.read().expect("should not be poisoned");
+        #[cfg(not(feature = "std"))]
+        let hooks = self.inner.read();
+
+        f(hooks.as_ref())
+    }
+
+    /// Divide frames, this looks a every strain and checks and finds the underlying variants
+    /// Those variants then are extracted, meaning:
+    ///
+    /// ```text
+    /// [A, B (Error), C, D (Error)]
+    /// ```
+    ///
+    /// turns into:
+    ///
+    /// ```text
+    /// [A, B (Error)]
+    /// [A, B, C, D (Error)]
+    /// ```
+    fn divide_frames<'a>(
+        &self,
+        frames: impl IntoIterator<Item = Vec<&'a Frame>>,
+    ) -> impl IntoIterator<Item = Vec<&'a Frame>> {
+        let ids: BTreeSet<_> = self.read_hooks_with(|inner| {
+            inner.map_or_else(BTreeSet::new, |inner| inner.keys().copied().collect())
+        });
+
+        frames.into_iter().flat_map(move |path| {
+            let mut div = vec![];
+            let mut walked = vec![];
+
+            for frame in path {
+                walked.push(frame);
+
+                if ids.contains(&Frame::type_id(frame)) {
+                    div.push(walked.clone());
+                }
+            }
+
+            div
+        })
+    }
+
+    fn serialize_report<S: Serializer>(
+        &self,
+        report: &Report<impl Context>,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error> {
+        self.init();
+
+        let frames = FrameSplitIterator::new(report);
+        let frames = self.divide_frames(frames);
+
+        serializer.collect_seq(frames.into_iter().filter_map(|stack| {
+            let last = stack.last()?;
+            let type_id = Frame::type_id(last);
+
+            // TODO: potentially we'd want to move this out, depending on performance
+            //  but this is an RwLock, which means that usually speed should not be of concern.
+            let hook: HookFn =
+                self.read_hooks_with(|hooks| hooks.and_then(|hooks| hooks.get(&type_id).copied()))?;
+
+            hook(stack.as_slice())
+        }))
+    }
+}
+
+static HOOKS: Hooks = Hooks::new();
+
+#[macro_export]
+macro_rules! register {
+    ($($ty:ident),+) => {
+        $crate::error::__private::register(&[$($crate::error::Hook::new::<$ty>(),)*])
+    };
+}
```

Comment:
What are the advantages of `register![a, b, c]` over `register(a); register(b); register(c);`?

Reply:
Mainly that it is more concise, we work on the type itself, meaning one would need to do `register::<A>();` instead, this might be confusing for some.

It is probably also quite common to initialize them all at once, so I thought it'd be beneficial to have less syntactic noise during init.